src/Tools/Auditor/Report.dfy

```diff
@@ -0,0 +1,125 @@
+
+  datatype Report =
+    Report(assumptions: seq<Assumption>)
+
+  const EmptyReport := Report([])
+
+  //// Tag categorization ////
+
+  predicate method IsAssumption(ts: set<Tag>) {
+    || (IsSubsetType in ts && HasNoWitness in ts)
+    || HasAxiomAttribute in ts
+    || (&& IsCallable in ts
+        && (|| (HasEnsuresClause in ts && (HasNoBody in ts || HasExternAttribute in ts))
+            || HasAssumeInBody in ts))
+  }
+
+  predicate method IsExplicitAssumption(ts: set<Tag>) {
+    HasAxiomAttribute in ts || HasAssumeInBody in ts
+  }
+
+  //// Report rendering ////
+
+  function method BoolYN(b: bool): string {
+    if b then "Y" else "N"
+  }
+
+  function method MaybeElt<T>(b: bool, elt: T): seq<T> {
+     if b then [elt] else []
+  }
+
+  // TODO: improve these descriptions
+  function method AssumptionDescription(ts: set<Tag>): seq<(string, string)> {
+    MaybeElt(IsCallable in ts && HasNoBody in ts && IsGhost in ts,
+      ("Function or lemma has no body.",
+       "Provide a body.")) +
+    MaybeElt(IsCallable in ts && HasNoBody in ts && !(IsGhost in ts),
+      ("Callable definition has no body.",
+       "Provide a body.")) +
+    MaybeElt(HasExternAttribute in ts && HasRequiresClause in ts,
+      ("Extern symbol with precondition.",
+       "Extensively test client code.")) +
+    MaybeElt(HasExternAttribute in ts && HasEnsuresClause in ts,
+      ("Extern symbol with postcondition.",
+       "Provide a model or a test case, or both.")) +
+    MaybeElt(IsSubsetType in ts && HasNoWitness in ts,
+      ("Subset type has no witness and could be empty.",
+       "Provide a witness.")) +
+    MaybeElt(HasAxiomAttribute in ts,
+      ("Has explicit `{:axiom}` attribute.",
+       "Attempt to provide a proof or model.")) +
+    MaybeElt(HasAssumeInBody in ts,
+      ("Has `assume` statement in body.",
+      "Try to replace with `assert` and prove."))
+  }
+
+  lemma AllAssumptionsDescribed(ts: set<Tag>)
+    requires IsAssumption(ts)
+    ensures |AssumptionDescription(ts)| > 0
+    {
+    }
+
+  function method RenderAssumptionMarkdown(a: Assumption): string {
+    var descs := AssumptionDescription(a.tags);
+    var issues := Map( (desc: (string, string)) => desc.0, descs);
+    var mitigations := Map( (desc: (string, string)) => desc.1, descs);
+    var cells :=
+      [ a.name
+      , BoolYN(!(IsGhost in a.tags))
+      , BoolYN(IsExplicitAssumption(a.tags))
+      , BoolYN(HasExternAttribute in a.tags)
+      , Flatten(Seq.Interleave("<br>", issues))
+      , Flatten(Seq.Interleave("<br>", mitigations))
+      ];
+    "| " + Flatten(Seq.Interleave(" | ", cells)) + " |"
+  }
+
+  function method RenderAuditReportMarkdown(r: Report): string {
+    var header :=
+      "|Name|Compiled|Explicit Assumption|Extern|Issue|Mitigation|\n" +
+      "|----|--------|-------------------|------|-----|----------|\n";
+    FoldL((s, a) => s + RenderAssumptionMarkdown(a) + "\n", header, r.assumptions)
+  }
```

Comment:
How readable will this be on the command line?  I would have grouped the report by "kind" of issue, followed by a small description, and then reported each "violation" like an error message, with a position and a name:

```
# High severity

[unimplemented] 5 missing implementations
  When a lemma, function, or method does not have a body, Dafny will assume its pre- and postconditions without proving them.  To resolve this issue, either provide a body or annotate the function as `{:extern}` (for executable code) or `{:axiom}` (for ghost code).

  foo.dfy:17: FooFunction
  bar.dfy:422: BarMethod

[naked-assume] 3 assumed properties
  `assume`s are not verified statically nor checked at run time, making them unsafe.  To resolve this issue, use `aasert` to check properties statically, `expect` to check them at run time (in methods only), or mark necessary assumptions that have been carefully reviewed with `{:axiom}`

  blah.dfy:17: assume foo
  blah.dfy:19: assume bar
  blah.dfy:22: assume baz

# Low severity

[axiomatized-assume] 2 assumed properties marked with `{:axiom}`
  `assume` is unsafe, but the following assumes were marked `{:axiom}` and are thus reported at low severity:

  …
```

Reply:
Ah, one thing I didn't mention in this PR so far: the goal is to have it render as HTML via `pandoc` with sortable columns. I have that working but haven't checked it into the repo so far. (And to do that we'll want to write it to a file, anyway.) Maybe having both Markdown and human-readable versions would be good, though.

Reply:
The plugin can now generate either HTML or Markdown in a file, and I'm about to add human-readable text for the console (or a file).